### PR TITLE
ESTs for policy sets

### PR DESCRIPTION
**Pull request by aaronjeline**
## Description of changes
Adds and EST for policy sets

## Issue #, if available
#549 
## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [X] A backwards-incompatible change requiring a major version bump to `cedar-policy` 

I confirm that this PR (choose one, and delete the other options):

- [x] Updates the "Unreleased" section of the CHANGELOG with a description of my change (required for major/minor version bumps).

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- [X] Does not require updates because my change does not impact the Cedar formal model or DRT infrastructure.

### CODE REVIEW

**Comment by aaronjeline** [comment]
Semver check is failing because `serde_json::Error` doesn't implement `UnwindSafe`, causing `PolicySetError` to no longer implement `UnwindSafe`. Not clear how much we care about this, should we:
A) Don't care, keep the rich error type
B) Convert `serde_json::Error` to a string, preserving the `UnwindSafe`

**Comment by aaronjeline** [comment]
> Semver check is failing because `serde_json::Error` doesn't implement `UnwindSafe`, causing `PolicySetError` to no longer implement `UnwindSafe`. Not clear how much we care about this, should we: A) Don't care, keep the rich error type B) Convert `serde_json::Error` to a string, preserving the `UnwindSafe`

I think we're fine with this break

File: cedar-policy-core/src/entities/json/value.rs [hunk 1]
```diff
@@ -726,17 +727,40 @@ impl<'e> ValueParser<'e> {
     }
 }
 
+/// A (optional) static context for deserialization of entity uids
+/// This is useful when for plumbing reasons, we can't get the appopriate values into the dynamic
```
Comment (andrewmwells-amazon) [inline cedar-policy-core/src/entities/json/value.rs:731]:
```suggestion
/// This is useful when, for plumbing reasons, we can't get the appopriate values into the dynamic
```

File: cedar-policy-core/src/est/err.rs [hunk 2]
```diff
@@ -80,6 +81,14 @@ pub enum FromJsonError {
     #[error("invalid entity type: {0}")]
     #[diagnostic(transparent)]
     InvalidEntityType(ParseErrors),
+    /// Error reported when a policy set has duplicate ids
+    #[error("Error creating policy set: {0}")]
+    #[diagnostic(transparent)]
+    PolicySet(#[from] PolicySetError),
+    /// Error reported when attempting to create an invalid template-link
```
Comment (andrewmwells-amazon) [inline cedar-policy-core/src/est/err.rs:88]:
```suggestion
    /// Error reported when attempting to create a template-link
```

Comment (andrewmwells-amazon) [reply cedar-policy-core/src/est/err.rs:88]: Presumably they were attempting to create a valid template-link 😄

File: cedar-policy/src/api.rs [hunk 3]
```diff
@@ -2622,6 +2704,51 @@ impl std::fmt::Display for PolicySet {
...
+            Either::Left(PolicyEntry {
+                id: id.0,
+                policy: est,
+            })
+        }),
+    }
+}
+
+/// Like [`itertools::Itertools::partion_map`], but accepts a function that can fail.
+/// The first invocation of `f` that fails causes the whole computation fails
```
Comment (andrewmwells-amazon) [inline cedar-policy/src/api.rs:2734]:
```suggestion
/// The first invocation of `f` that fails causes the whole computation to fail
```

File: cedar-policy/src/api.rs [hunk 4]
```diff
@@ -2622,6 +2704,51 @@ impl std::fmt::Display for PolicySet {
     }
 }
 
+/// Given a [`PolicyId`] and a [`Policy`], determine if the policy represents a static policy or a
+/// link
+fn parition_policies(
```
Comment (andrewmwells-amazon) [inline cedar-policy/src/api.rs:2709]: I don't like the name `partition`. Maybe something like `is_static_or_link`?

File: cedar-policy/src/api.rs [hunk 5]
```diff
@@ -2622,6 +2704,51 @@ impl std::fmt::Display for PolicySet {
...
+                id: id.0,
+                policy: est,
+            })
+        }),
+    }
+}
+
+/// Like [`itertools::Itertools::partion_map`], but accepts a function that can fail.
+/// The first invocation of `f` that fails causes the whole computation fails
+fn fold_parition<T, A, B, E>(
```
Comment (andrewmwells-amazon) [inline cedar-policy/src/api.rs:2735]:
```suggestion
fn fold_partition<T, A, B, E>(
```

Comment (andrewmwells-amazon) [reply cedar-policy/src/api.rs:2735]: And the same throughout

**Review by andrewmwells-amazon — CHANGES_REQUESTED** [review]

**Review by shaobo-he-aws — APPROVED** [review]

**Comment by shaobo-he-aws** [comment]
LGTM. I think we will also need to update the docs because policy links now have EST representations.

File: cedar-policy/src/api.rs [hunk 6]
```diff
@@ -2269,6 +2271,17 @@ pub enum PolicySetError {
     /// Error when removing a link that is not a link
     #[error("unable to unlink `{0}` because it is not a link")]
     UnlinkLinkNotLinkError(PolicyId),
+    /// Error when converting from EST
+    #[error("Error deserializing an EST policy/template: {0}")]
+    #[diagnostic(transparent)]
+    FromEST(#[from] cedar_policy_core::est::FromJsonError),
+    /// Error when covnerting to EST
+    #[error("Error serialzing a policy to the EST format: {0}")]
```
Comment (cdisselkoen) [inline cedar-policy/src/api.rs:2279]:
```suggestion
    #[error("Error serializing a policy to the EST format: {0}")]
```

File: cedar-policy/src/api.rs [hunk 7]
```diff
@@ -2341,6 +2354,73 @@ impl FromStr for PolicySet {
...
+        }
+
+        Ok(pset)
+    }
+
+    /// Deserialize the [`PolicySet`] from a JSON string
+    pub fn from_json_str(src: impl AsRef<str>) -> Result<Self, PolicySetError> {
+        let est: est::PolicySet = serde_json::from_str(src.as_ref())?;
+        Self::from_est(est)
+    }
```
Comment (cdisselkoen) [inline cedar-policy/src/api.rs:2392]: I think in other places we have the full trifecta of json constructors -- from value, str, and Read

File: cedar-policy/src/api.rs [hunk 8]
```diff
@@ -2622,6 +2702,51 @@ impl std::fmt::Display for PolicySet {
...
+/// link
+fn is_static_or_link(
+    (id, policy): (PolicyId, Policy),
+) -> Result<Either<est::PolicyEntry, Link>, PolicyToJsonError> {
+    let env = policy
+        .ast
+        .env()
+        .iter()
+        .map(|(id, euid)| (*id, euid.clone()))
+        .collect();
```
Comment (cdisselkoen) [inline cedar-policy/src/api.rs:2715]: nit: we only need `env` in the `Some(template_id)` case, so we can move this computation into that branch

File: cedar-policy/src/api.rs [hunk 9]
```diff
@@ -2622,6 +2702,51 @@ impl std::fmt::Display for PolicySet {
...
+        None => policy.lossless.est().map(|est| {
+            Either::Left(PolicyEntry {
+                id: id.0,
+                policy: est,
+            })
+        }),
+    }
+}
+
+/// Like [`itertools::Itertools::partion_map`], but accepts a function that can fail.
```
Comment (cdisselkoen) [inline cedar-policy/src/api.rs:2731]:
```suggestion
/// Like [`itertools::Itertools::partition_map`], but accepts a function that can fail.
```

File: cedar-policy/src/api.rs [hunk 10]
```diff
@@ -4469,4 +4608,639 @@ mod test {
         list_out.sort();
         assert_eq!(list_out, &["admin", "alice"]);
     }
+
+    #[test]
+    fn test_parition_fold() {
```
Comment (cdisselkoen) [inline cedar-policy/src/api.rs:4613]:
```suggestion
    fn test_partition_fold() {
```

and elsewhere below

**Review by cdisselkoen — CHANGES_REQUESTED** [review]
Some of these more urgent than others.  Bunch of these are just comments/questions why things are a certain way

File: cedar-policy-core/src/entities/json/value.rs [hunk 11]
```diff
@@ -726,17 +727,40 @@ impl<'e> ValueParser<'e> {
...
-pub enum EntityUidJson {
+pub enum EntityUidJson<Context = RuntimeContext> {
     /// This was removed in 3.0 and is only here for generating nice error messages.
     ExplicitExprEscape {
         /// Contents are ignored.
         __expr: String,
+        /// Phantom value for the `Context` type parameter
+        #[serde(skip)]
+        context: std::marker::PhantomData<Context>,
     },
```
[the remark this answers is not in the document]
Comment (aaronjeline) [reply cedar-policy-core/src/entities/json/value.rs:764]: Had to go in one of the cases, just picked one

File: cedar-policy-core/src/entities/json/value.rs [hunk 12]
```diff
@@ -765,8 +813,9 @@ impl EntityUidJson {
     /// Convert this `EntityUidJson` into an `EntityUID`
     pub fn into_euid(
         self,
-        ctx: impl Fn() -> JsonDeserializationErrorContext + Clone,
+        dynamic_ctx: impl Fn() -> JsonDeserializationErrorContext + Clone,
     ) -> Result<EntityUID, JsonDeserializationError> {
+        let ctx = || C::static_context().unwrap_or_else(&dynamic_ctx);
```
[the remark this answers is not in the document]
Comment (aaronjeline) [reply cedar-policy-core/src/entities/json/value.rs:818]: Dynamic context will always be present, (at least with the current type signature)

File: cedar-policy-core/src/entities/json/value.rs [hunk 13]
```diff
@@ -726,17 +727,40 @@ impl<'e> ValueParser<'e> {
...
 }
 
+/// A (optional) static context for deserialization of entity uids
+/// This is useful when, for plumbing reasons, we can't get the appopriate values into the dynamic
+/// context. Primary use case is in the [`DeserializeAs`] trait.
+pub trait DeserializationContext {
+    /// Access the (optional) static context.
+    /// If returns [`None`], use the dynamic context.
+    fn static_context() -> Option<JsonDeserializationErrorContext>;
+}
```
[the remark this answers is not in the document]
Comment (aaronjeline) [reply cedar-policy-core/src/entities/json/value.rs:737]: It's unclear to me how you'd thread that argument through though

**Review by shaobo-he-aws — APPROVED** [review]

File: cedar-policy-core/src/entities/json/value.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (cdisselkoen) [reply cedar-policy-core/src/entities/json/value.rs:737]: You're probably right, I was just musing

File: cedar-policy-core/src/entities/json/value.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (john-h-kastner-aws) [reply cedar-policy-core/src/entities/json/value.rs:818]: If we always have a dynamic context, what's the purpose of a static context?

[the remark this answers is not in the document]
Comment (aaronjeline) [reply cedar-policy-core/src/entities/json/value.rs:818]: Sometimes the dynamic context can't be set correct (for example in FromStr impls) so we need to get the correct context in there

File: cedar-policy/src/api.rs [hunk 14]
```diff
@@ -2269,6 +2270,17 @@ pub enum PolicySetError {
     /// Error when removing a link that is not a link
     #[error("unable to unlink `{0}` because it is not a link")]
     UnlinkLinkNotLinkError(PolicyId),
+    /// Error when converting from EST
+    #[error("Error deserializing an EST policy/template: {0}")]
+    #[diagnostic(transparent)]
+    FromEST(#[from] cedar_policy_core::est::FromJsonError),
+    /// Error when covnerting to EST
```
Comment (cdisselkoen) [inline cedar-policy/src/api.rs:2277]:
```suggestion
    /// Error when converting to EST
```

**Review by cdisselkoen — APPROVED** [review]

File: cedar-policy-core/src/est/policy_set.rs [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+#[serde_as]
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct Link {
+    /// The id of the link
+    pub id: PolicyID,
+    /// The id of the template
+    pub target: PolicyID,
+    /// The mapping between slots and entity uids
+    #[serde_as(as = "serde_with::MapPreventDuplicates<_,EntityUidJson<SlotEnvContext>>")]
+    pub env: HashMap<SlotId, EntityUID>,
```
[the remark this answers is not in the document]
Comment (aaronjeline) [reply cedar-policy-core/src/est/policy_set.rs:58]: Yeah I agree

File: cedar-policy/CHANGELOG.md [hunk 16]
```diff
@@ -13,6 +13,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - `wasm` Cargo feature for targeting Wasm
 - `Entity::into_inner` (resolving #636)
 - `Entities::into_iter` (resolving #680)
+- Adds an EST representation for Policy Sets (#783, resolving #549)
```
Comment (cdisselkoen) [inline cedar-policy/CHANGELOG.md:16]: nit: maybe list the specific methods being added.  People don't necessarily know what "EST" is, that is internal terminology

File: cedar-policy/src/api.rs [hunk 17]
```diff
@@ -2269,6 +2270,17 @@ pub enum PolicySetError {
...
     #[error("unable to unlink `{0}` because it is not a link")]
     UnlinkLinkNotLinkError(PolicyId),
+    /// Error when converting from EST
+    #[error("Error deserializing an EST policy/template: {0}")]
+    #[diagnostic(transparent)]
+    FromEST(#[from] cedar_policy_core::est::FromJsonError),
+    /// Error when converting to EST
+    #[error("Error serializing a policy to the EST format: {0}")]
+    #[diagnostic(transparent)]
+    ToEST(#[from] PolicyToJsonError),
```
Comment (cdisselkoen) [inline cedar-policy/src/api.rs:2280]: Similarly here we can avoid using the "EST" internal terminology and just say `FromJson` and "Error deserializing a JSON policy/template" and "Error serializing a policy to the JSON format" etc

**Review by cdisselkoen — APPROVED** [review]

**Pull request merged**